torch/distributed/_composable/fsdp/fully_shard.py

```diff
@@ -0,0 +1,48 @@
+from typing import Any
+
+import typing_extensions
+
+import torch.nn as nn
+
+from torch.distributed._composable import contract
+from torch.distributed._composable_state import _insert_module_state
+
+
+# The decorator adds a state object to `module` that can be accessed via
+# `fully_shard.state(module)`. The state object and module are 1:1.
+@contract()
+def fully_shard(
+    module: nn.Module,
+):
+    if isinstance(module, (nn.ModuleList, nn.ModuleDict)):
+        raise ValueError(
+            f"fully_shard does not support containers that do not implement forward: {module}"
+        )
+    state = fully_shard.state(module)
+    _insert_module_state(module, state)
+    # Place FSDP leftmost for highest priority in the method resolution order
+    cls = module.__class__
+    dct = {"__deepcopy__": unimplemented_deepcopy}
+    new_cls = type(f"FSDP{cls.__name__}", (FSDP, cls), dct)
```

Comment:
For my learning purpose - does the dynamic mixin approach offer more desirable properties for FSDP2 over the composition approach, or is it mostly a stylistic choice?

Reply:
To clarify, what is the composition approach?

Reply:
I meant comparing to the user module being a field of the FSDP module.

Reply:
Ah, got it!

We do not want FSDP to be an `nn.Module` wrapper since the addition of a prefix to the FQNs complicates state dict logic. Saving requires removing the prefix, and loading requires adding the prefix. Even worse, FSDP had to be aware of other APIs like AC:
https://github.com/pytorch/pytorch/blob/e33e88e5bcb6810c0dbd6b7516c2d07c71eef2a8/torch/distributed/fsdp/fully_sharded_data_parallel.py#L524-L531

The effort to avoid `nn.Module` wrappers at the end of 2022 H2 led to the `torch/distributed/_composable` folder, where we had versions of DDP/FSDP/AC that relied on `nn.Module` forward hooks instead. This was somewhat formalized with the `@contract` decorator that enforced that FQNs did not change before/after applying the API.

The issue then though was how to override or add methods. Without the dynamic mixin, we would need to monkey patch the methods. The mixin provides more structure and allows for `isinstance(module, FSDP)` checks.